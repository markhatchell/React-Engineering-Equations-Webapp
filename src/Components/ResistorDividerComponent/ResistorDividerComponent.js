import React from 'react';
import BaseEquation from '../BaseEquation';
import '../../Styles/Equations.css';
const { ResistorDivider } = require('js-electrical-engineering-equations');

class ResistorDividerComponent extends BaseEquation {

  constructor(props) {
    super(props);
    if (props.hasOwnProperty('onChangeTitle')) {
      props.onChangeTitle('Resistor Divider');
    }
    this.state = {
      inputVoltage: '',
      resistor1: '',
      resistor2: '',
      outputVoltage: '',
      lastCalculated: []
    }
  }

  render() {

    return (
      <div className="equationBox">
        <h3>Resistor Divider</h3>
        <div className="equationBoxInner">
          <p>Supply at least three of the values to calculate the fourth.</p>
          <p>The base equation for a resistor divider network is <br /> Vo = (Vs * R2) / (R1 + R2)</p>
          <div className="tableContainer">
            <form onSubmit={(e) => this.handleSubmit(e)}>
              <table className="table">
                <tbody>
                <tr>
                  <td align="left">
                    <label htmlFor="omhslaw-I">Input Voltage</label>
                  </td>
                  <td>
                    <input id="inputVoltage" value={this.state.inputVoltage} disabled={this.wasLastCalculated('inputVoltage')}
                           autoComplete="off"
                           tabIndex="1"
                           onChange={(e) => this.collectFloatValueFor('inputVoltage', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <label htmlFor="omhslaw-V">Resistor 1 (Ohms)</label>
                  </td>
                  <td>
                    <input id="resistor1" value={this.state.resistor1} disabled={this.wasLastCalculated('resistor1')}
                           autoComplete="off"
                           tabIndex="2"
                           onChange={(e) => this.collectFloatValueFor('resistor1', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <label htmlFor="omhslaw-R">Resistor 2 (Ohms)</label>
                  </td>
                  <td>
                    <input id="resistor2" value={this.state.resistor2} disabled={this.wasLastCalculated('resistor2')}
                           autoComplete="off"
                           tabIndex="3"
                           onChange={(e) => this.collectFloatValueFor('resistor2', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <label htmlFor="omhslaw-W">Output Voltage</label>
                  </td>
                  <td>
                    <input id="outputVoltage" value={this.state.outputVoltage} disabled={this.wasLastCalculated('outputVoltage')}
                           autoComplete="off"
                           tabIndex="4"
                           onChange={(e) => this.collectFloatValueFor('outputVoltage', e)}/>
                  </td>
                </tr>
                <tr>
                  <td align="left">
                    <button type="button" tabIndex="6" onClick={() => this.handleClear()}>Clear</button>
                  </td>
                  <td align="right">
                    <button type="submit" tabIndex="5">Calculate</button>
                  </td>
                </tr>
                </tbody>
              </table>
            </form>
          </div>
        </div>
      </div>
    );
  }

  calculateOtherValues() {
    const {inputVoltage, resistor1, resistor2, outputVoltage} = this.state;

    if ((this.wasLastCalculated('outputVoltage')) || (inputVoltage && resistor1 && resistor2 && outputVoltage === '')) {
      this.setState(
        Object.assign(
          {},
          this.state,
          ResistorDivider.calcOutputVoltage(parseFloat(inputVoltage), parseFloat(resistor1), parseFloat(resistor2)),
          {
            lastCalculated: ['outputVoltage']
          }
        )
      );
    } else if ((this.wasLastCalculated('resistor1')) || (inputVoltage && resistor2 && outputVoltage && resistor1 === '')) {
      this.setState(
        Object.assign(
          {},
          this.state,
          ResistorDivider.calcResistor1(parseFloat(inputVoltage), parseFloat(outputVoltage), parseFloat(resistor2)),
          {
            lastCalculated: ['resistor1']
          }
        )
      );
    } else if ((this.wasLastCalculated('resistor2')) || (inputVoltage && resistor1 && outputVoltage && resistor2 === '')) {
      this.setState(
        Object.assign(
          {},
          this.state,
          ResistorDivider.calcResistor2(parseFloat(inputVoltage), parseFloat(outputVoltage), parseFloat(resistor1)),
          {
            lastCalculated: ['resistor2']
          }
        )
      );
    } else if ((this.wasLastCalculated('inputVoltage')) || (outputVoltage && resistor1 && resistor2 && inputVoltage === '')) {
      this.setState(
        Object.assign(
          {},
          this.state,
          ResistorDivider.calcInputVoltage(parseFloat(outputVoltage), parseFloat(resistor1), parseFloat(resistor2)),
          {
            lastCalculated: ['inputVoltage']
          }
        )
      );
    }

  }

  countValues() {
    const values = [this.state.inputVoltage, this.state.resistor1, this.state.resistor2, this.state.outputVoltage];
    let valuesCount = 0;
    for (let i = 0; i < values.length; i++) {
      if (values[i] !== '') {
        valuesCount++;
      }
    }
    return valuesCount;
  }

  handleSubmit(e) {
    const {lastCalculated} = this.state;
    if (this.countValues() === 3 || lastCalculated !== []) {
      this.calculateOtherValues();
    } else {
      this.clear();
    }
    e.preventDefault();
    return false;
  }

  clear() {
    this.setState(
      Object.assign(
        {},
        this.state,
        {
          inputVoltage: '',
          resistor1: '',
          resistor2: '',
          outputVoltage: '',
          lastCalculated: ''
        }
      )
    );
  }


  handleClear() {
    this.clear();
  }

  wasLastCalculated(field) {
    const lastCalculated = this.state.lastCalculated;
    return lastCalculated.indexOf(field) >= 0;

  }

}


export default ResistorDividerComponent;