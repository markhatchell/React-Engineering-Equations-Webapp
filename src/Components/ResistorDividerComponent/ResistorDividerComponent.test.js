import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ResistorDividerComponent from "./ResistorDividerComponent";

configure({ adapter: new Adapter() });

const dom = renderer.create(
  <ResistorDividerComponent />,
);

describe('ResistorDividerComponent', () => {

  test('snapshot matches', () => {
    let tree = dom.toJSON();
    expect(tree).toMatchSnapshot();
  });

  const component = shallow(<ResistorDividerComponent />);
  const instance = component.instance();



});
