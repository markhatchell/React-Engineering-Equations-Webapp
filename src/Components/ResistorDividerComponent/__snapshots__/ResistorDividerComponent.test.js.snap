// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResistorDividerComponent snapshot matches 1`] = `
<div
  className="equationBox"
>
  <h3>
    Resistor Divider
  </h3>
  <div
    className="equationBoxInner"
  >
    <p>
      Supply at least three of the values to calculate the fourth.
    </p>
    <p>
      The base equation for a resistor divider network is 
      <br />
       Vo = (Vs * R2) / (R1 + R2)
    </p>
    <div
      className="tableContainer"
    >
      <form
        onSubmit={[Function]}
      >
        <table
          className="table"
        >
          <tbody>
            <tr>
              <td
                align="left"
              >
                <label
                  htmlFor="omhslaw-I"
                >
                  Input Voltage
                </label>
              </td>
              <td>
                <input
                  autoComplete="off"
                  disabled={false}
                  id="inputVoltage"
                  onChange={[Function]}
                  tabIndex="1"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td
                align="left"
              >
                <label
                  htmlFor="omhslaw-V"
                >
                  Resistor 1 (Ohms)
                </label>
              </td>
              <td>
                <input
                  autoComplete="off"
                  disabled={false}
                  id="resistor1"
                  onChange={[Function]}
                  tabIndex="2"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td
                align="left"
              >
                <label
                  htmlFor="omhslaw-R"
                >
                  Resistor 2 (Ohms)
                </label>
              </td>
              <td>
                <input
                  autoComplete="off"
                  disabled={false}
                  id="resistor2"
                  onChange={[Function]}
                  tabIndex="3"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td
                align="left"
              >
                <label
                  htmlFor="omhslaw-W"
                >
                  Output Voltage
                </label>
              </td>
              <td>
                <input
                  autoComplete="off"
                  disabled={false}
                  id="outputVoltage"
                  onChange={[Function]}
                  tabIndex="4"
                  value=""
                />
              </td>
            </tr>
            <tr>
              <td
                align="left"
              >
                <button
                  onClick={[Function]}
                  tabIndex="6"
                  type="button"
                >
                  Clear
                </button>
              </td>
              <td
                align="right"
              >
                <button
                  tabIndex="5"
                  type="submit"
                >
                  Calculate
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  </div>
</div>
`;
